compiler/rustc_typeck/src/check/check.rs

```diff
@@ -987,12 +991,37 @@ fn check_impl_items_against_trait<'tcx>(
             if !is_implemented && tcx.impl_defaultness(impl_id).is_final() {
                 missing_items.push(tcx.associated_item(trait_item_id));
             }
+
+            if let Some(required_items) = &must_implement_one_of {
+                // true if this item is specifically implemented in this impl
```

Comment:
```suggestion
                // true if this item is specifically implemented in this impl ≠ a default implementation 
```

Speaking of which: how does this interact with specialization / partial impls?

```rust
#[rustc_must_implement_one_of(foo, bar)]
trait Trait {
    fn foo() { Self::bar() }
    fn bar() { Self::foo() }
}

default
impl Trait for () {
    fn foo() {}
}

impl Trait for () {}
```

Reply:
This example doesn't compile even without `rustc_must_implement_one_of` though...
```text
error[E0119]: conflicting implementations of trait `Trait` for type `()`
  --> ./what.rs:15:1
   |
10 | / default
11 | | impl Trait for () {
12 | |     fn foo() {}
13 | | }
   | |_- first implementation here
14 |
15 |   impl Trait for () {}
   |   ^^^^^^^^^^^^^^^^^ conflicting implementation for `()`
```

Reply:
Argh, I hate `default impl` not meaning `partial impl` but `partial impl { default members }`. Okay, replace `default impl Trait for ()` with `default impl<T> Trait for T`

Reply:
So
- `default impl<T> Trait for T { fn foo() {} }` & `impl Trait for () {}` compiles
- `default impl<T> Trait for T {}` errors

I'm honestly very confused about what `default` impls mean, so not sure if this is correct 😅 

<details><summary>Snippets</summary>
<p>

```rust
#![feature(specialization)]
#![feature(rustc_attrs)]

#[rustc_must_implement_one_of(foo, bar)]
trait Trait {
    fn foo() { Self::bar() }
    fn bar() { Self::foo() }
}

default impl<T> Trait for T { 
    fn foo() {} 
}

impl Trait for () {}

fn main() {}
```
```rust
#![feature(specialization)]
#![feature(rustc_attrs)]

#[rustc_must_implement_one_of(foo, bar)]
trait Trait {
    fn foo() { Self::bar() }
    fn bar() { Self::foo() }
}

default impl<T> Trait for T {}

fn main() {}
```
```text
error[E0046]: not all trait items implemented, missing one of: `foo`, `bar`
  --> ./what.rs:10:1
   |
10 | default impl<T> Trait for T {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing one of `foo`, `bar` in implementation
   |
note: required because of this annotation
  --> ./what.rs:4:1
   |
4  | #[rustc_must_implement_one_of(foo, bar)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
</p>
</details>

Reply:
> what `default` impls means

<details><summary>My view on the matter</summary>

So, `default impl … { items… }` means `partial impl … { default items… }`, as in one is then writing part of a trait (from 0 to _all_ associated items), but for the trait itself. And I personally see `partial impl`, _alone_, as sugar for super-traits.

Let's take `Any` upcasting as an example:

```rust
trait SubAny : Any {
    fn as_any(&self) -> &dyn Any;
}
```

Problem: implementors of `SubAny` would need to feature an impl of `as_any`.

```rust
trait SubAny : Any {
    fn as_any(&self) -> &dyn Any { self }
}
```

Problem: `Self : Sized` does not hold.

```rust
trait SubAny : Any {
    fn as_any(&self) -> &dyn Any
    where Self : Sized,
    { self }
}
```

Problem: `&dyn SubAny` can't use `as_any`.

```rust
trait SubAny : Any + AsAny {}
trait AsAny : Any {
    fn as_any(&self) -> &dyn Any;
}
impl<T : /* Sized + */ Any> AsAny for T {
    fn as_any(&self) -> &dyn Any { self }
}
```

This finally works, but is cumbersome, plus the `AsAny` trait does not read as well in the docs to let the user know of a `.as_any()` method available to it.

With `partial impl`s, however, we can both feature a method with no trait-def impl / body (where we couldn't because of lack of `Sized`) and yet not force implementors to provide, each time, the same body again and again:

```rust
trait SubAny : Any {
    fn as_any(&self) -> &dyn Any;
}

partial
impl<T : /* Sized + */ Any> SubAny for T {
    fn as_any(&self) -> &dyn Any { self }
}
```

___

</details>

Now, from here, your challenge would be to somehow make it if you are inside a `default impl`, not error _there_ (since even if some methods are implemented, the trait itself isn't and thus, its methods aren't callable, so there is no risk of infinite recursion of, _e.g._, `eq` calling `ne` calling `eq`, _etc._).

Given the body of the function, it may be as simple as disabling your check if `tcx.impl_defaultness(impl_id).is_final()` does not hold, but take that suggestion with a big grain of salt 😅

Reply:
Personally, I think it's fine to address any issues with `default` in a follow-up PR. Specialization is a long ways from stabilization, so I don't think it needs to block this PR